src/models/thread.ts

```diff
@@ -504,8 +504,42 @@ export class Thread extends ReadReceipt<EmittedEvents, EventHandlerMap> {
         throw new Error("Unsupported function on the thread model");
     }
 
+    public getEventReadUpTo(userId: string, ignoreSynthesized?: boolean): string | null {
+        const isCurrentUser = userId === this.client.getUserId();
+        if (isCurrentUser && this.lastReply()) {
+            // If a thread last activity is prior the first read receipt sent in a thread
+            // we want to consider that this thread has been read
+            const beforeFirstThreadedReceipt = (this.lastReply()?.getTs() ?? 0) < this.room.oldestThreadedReceiptTs;
+            if (beforeFirstThreadedReceipt) {
+                return this.timeline.at(-1)?.getId() ?? null;
+            }
+        }
+
+        const readUpToId = super.getEventReadUpTo(userId, ignoreSynthesized);
+
+        // Checking whether the unthreaded read receipt for that user is more recent
+        // than the read receipt inside that thread.
+        if (isCurrentUser && this.lastReply()) {
+            const unthreadedReceiptTs = this.room.unthreadedReceipts.get(userId)?.ts ?? Infinity;
```

Comment:
There's no reason to iterate the timeline if there's no unthreaded receipt:

```suggestion
            if (!this.room.unthreadedReceipts.has(userId)) {
                return readUpToId;
            }
            const unthreadedReceiptTs = this.room.unthreadedReceipts.get(userId).ts;
```